base-containers/base/inginious_container_api/run_student.py

```diff
@@ -221,6 +154,165 @@ def _hack_signals(receive_signal):
             except:
                 pass
 
+
 async def _send_intern_message(send_socket, msg):
+    """ Send a signal to the grading_container main process"""
     send_socket.send(msgpack.dumps(msg, use_bin_type=True))
     send_socket.recv()
+
+
+def create_student_socket(both_dockers):
+    """ Create a socket for the grading - student containers communication. Only used when both are using docker runtimes """
+    # creates a placeholder for the socket
+    DIR = "/sockets/"
+    _, path = tempfile.mkstemp('', 'p', DIR)
+
+    # Gets the socket id
+    socket_id = os.path.split(path)[-1]
+    socket_path = os.path.join(DIR, socket_id + ".sock")
+
+    if both_dockers:
+        # Start the socket
+        server = socket.socket(socket.AF_UNIX)
+        try:
+            os.unlink(socket_path)
+        except OSError:
+            if os.path.exists(socket_path):
+                raise
+        server.bind(socket_path)
+        server.listen(0)
+        return server, socket_id, socket_path, path
+    else:
+        return None, socket_id, socket_path, path
+
+
+def start_student_container(container, time_limit, hard_time_limit, memory_limit, share_network, socket_id, ssh, run_as_root):
+    """ Ask the docker agent to create the student container """
+    context = zmq.Context()
+    zmq_socket = context.socket(zmq.REQ)
+    zmq_socket.connect("ipc:///sockets/main.sock")
+    zmq_socket.send(msgpack.dumps({"type": "run_student", "environment": container,
+                                   "time_limit": time_limit, "hard_time_limit": hard_time_limit,
+                                   "memory_limit": memory_limit, "share_network": share_network,
+                                   "socket_id": socket_id, "ssh": ssh, "run_as_root": run_as_root},
+                                  use_bin_type=True))
+    # Check if the container was correctly started
+    message = msgpack.loads(zmq_socket.recv(), use_list=False, strict_map_key=False)
+    assert message["type"] == "run_student_started"
+    student_container_id = message["container_id"]
+    return zmq_socket, student_container_id
+
+
+def send_initial_command(socket_id, server, stdin, stdout, stderr, zmq_socket, student_container_id, cmd, teardown_script, working_dir, ssh, user, both_dockers):
+    """ Send the commands (aka: student code) to be run in the student container """
+    if both_dockers:
+        # Serve one and only one connection
+        connection, addr = server.accept()
+
+        # _run_student_intern should say hello
+        datagram = connection.recv(1)
+        assert datagram == b'H'
+
+        # send the fds and the command/workdir directly to student_container
+        connection.sendmsg([b'S'], [(socket.SOL_SOCKET, socket.SCM_RIGHTS, array.array("i", [stdin, stdout, stderr]))])
+        connection.send(msgpack.dumps(
+            {"type": "run_student_command", "student_container_id": student_container_id, "command": cmd,
```

Comment:
Indeed, the socket_id is not sent because you are looking at the "both_dockers" part and the socket_id is not required since the communications are direct grading - student without going through the docker agent.

Reply:
OK this is not required in this case, but the field is always fetched in all the message treatment routines I see with the ``[]`` operator, without any additional conditions. Where's the treatment routine for this specific message then ? Or is the field added by some code in the way ?

Reply:
I renamed the message to have a different name between only_dockers communications and not only_dockers ones.
Also added a verification of the message "type" on the student_container side.